src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+
+            ChainedHeader bestTip = this.chainState.ConsensusTip;
+            ConnectedHeaders connectedHeaders = new ConnectedHeaders() {Consumed = latestNewHeader};
+
+            if (latestNewHeader.ChainWork > bestTip.ChainWork)
+            {
+                this.logger.LogDebug("Chained header '{0}' is the tip of a chain with more work than our current consensus tip.", latestNewHeader);
+
+                connectedHeaders = this.MarkBetterChainAsRequired(latestNewHeader);
+            }
+
+            this.MarkTrustedChainAsAssumedValid(assumedValidHeader);
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// When a header is before the last checkpoint it will be asked to be marked as AssumedValid and we'll ask to download it.
+        /// </summary>
+        private ConnectedHeaders HandleCheckpointsHeader(ChainedHeader chainedHeader, ChainedHeader latestNewHeader, CheckpointInfo checkpoint)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}',{4}.{5}:'{6}')", nameof(chainedHeader), chainedHeader, nameof(latestNewHeader), latestNewHeader, nameof(checkpoint), nameof(checkpoint.Hash), checkpoint.Hash);
+
+            if (checkpoint.Hash != chainedHeader.HashBlock)
+            {
+                this.logger.LogDebug("Chained header '{0}' does not match checkpoint '{1}'.", chainedHeader, checkpoint.Hash);
+                this.logger.LogTrace("(-)[INVALID_HEADER_NOT_MATCHING_CHECKPOINT]");
+                throw new InvalidHeaderException();
+            }
+
+            ChainedHeader subchainTip = chainedHeader;
+            if (chainedHeader.Height == this.checkpoints.GetLastCheckpointHeight())
+                subchainTip = latestNewHeader;
+
+            ConnectedHeaders connectedHeaders = this.MarkBetterChainAsRequired(subchainTip);
+            this.MarkTrustedChainAsAssumedValid(chainedHeader);
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// Check whether a header is in one of the following states
+        /// <see cref="BlockDataAvailabilityState.BlockAvailable"/>, <see cref="BlockDataAvailabilityState.BlockRequired"/>.
+        /// </summary>
+        private bool HeaderWasNotRequested(ChainedHeader chainedHeader)
+        {
+            return !((chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockAvailable)
+                  || (chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockRequired));
+        }
+
+        /// <summary>
+        /// Check whether a header is in one of the following states
+        /// <see cref="ValidationState.AssumedValid"/>, <see cref="ValidationState.PartiallyValidated"/>, <see cref="ValidationState.FullyValidated"/>.
+        /// </summary>
+        private bool HeaderWasNotMarkedAsValidated(ChainedHeader chainedHeader)
```

Comment:
```
return !((chainedHeader.BlockValidationState == ValidationState.AssumedValid)
                  || (chainedHeader.BlockValidationState == ValidationState.PartiallyValidated)
                  || (chainedHeader.BlockValidationState == ValidationState.FullyValidated));
```

is the same as 

```
return chainedHeader.BlockValidationState < ValidationState.AssumedValid;
```

So maybe replace those and make this expression inline? Method is used only in a single place anyway

Reply:
I'm not sure that I agree because we have not defined such an ordering so maybe it seems better to have explicit enumeration.

Reply:
Ofc we can go with this change only if we explicitly define the values (in the enum definition do AssumedValid = 1, PartiallyValidated = 2 etc)

Also don't mind if we don't since we do this kinds of check just in a few places